crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs

```diff
@@ -60,42 +286,705 @@ declare_lint_rule! {
+                    }
+                    _ => {
+                        // Use fallback instead
+                    }
+                },
+                AnyJsBindingPattern::JsObjectBindingPattern(named_bindings) => {
+                    // const { ... } = await import(...)
+                    return self.visit_named_bindings(named_bindings);
+                }
+                AnyJsBindingPattern::JsArrayBindingPattern(_) => {
+                    // const [ ... ] = await import(...)
+                    //
+                    // Array binding patterns do not really make sense for an import,
+                    // so discard the additonal information and use fallback instead.
+                }
+            }
+        };
+
+        // We failed to find any additional context, and are therefore
+        // restricted to analyzing "import(...)" as a namespace import,
+        // because that what is returned by "import(...)".
+        //
+        // The diagnostic will be associated with "import('module-name')"
+        // instead of just "'module_name'" to indicate that not the
+        // imported module itself is forbidden, but the ways in which
+        // it can be imported are restricted.
+        self.visit_special_import_node(import_call.syntax(), Self::NAMESPACE_IMPORT_ALIAS)
+    }
+
+    fn get_context_for_import_call(
+        import_call: &JsImportCallExpression,
+    ) -> Option<AnyJsBindingPattern> {
+        let mut current = import_call.syntax().parent()?;
+
+        while current.kind() == JsSyntaxKind::JS_PARENTHESIZED_EXPRESSION {
+            // #1: const { ... } = (await **(import(""))**)
+            // #2: **(import(""))**.then(...)
+            current = current.parent()?;
+        }
+
+        if current.kind() == JsSyntaxKind::JS_AWAIT_EXPRESSION {
+            // #1: const { ... } = (**await (import(""))**)
+            current = current.parent()?;
+
+            while current.kind() == JsSyntaxKind::JS_PARENTHESIZED_EXPRESSION {
+                // #1: const { ... } = **(await (import("")))**
+                current = current.parent()?;
+            }
+        } else if current.kind() == JsSyntaxKind::JS_STATIC_MEMBER_EXPRESSION {
+            // #2: **(import("")).then**(...)
+            let static_member_expr = current.cast::<JsStaticMemberExpression>()?;
+            let member_name = static_member_expr.member().ok()?;
+            if member_name.as_js_name()?.text() != "then" {
+                return None;
+            }
+            current = static_member_expr.syntax().parent()?;
+
+            if current.kind() == JsSyntaxKind::JS_CALL_EXPRESSION {
+                // #2: **(import("")).then(...)**
+                let then_call_expr = current.cast::<JsCallExpression>()?;
+                let then_call_arg = then_call_expr
+                    .arguments()
+                    .ok()?
+                    .args()
+                    .into_iter()
```

Comment:
There's no need to use `into_iter` with our CST because internally, they behave as `iter`. We try to use `iter` instead

Reply:
No worries! The code is actually correct! However, `into_iter` comes from our internal implementation and architecture, that's why there's this weird behaviour 😅